File: samples/cpp/text2image/main.cpp
// Copyright (C) 2023-2024 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include "openvino/genai/image_generation/text2image_pipeline.hpp"

#include "imwrite.hpp"

int32_t main(int32_t argc, char* argv[]) try {
    OPENVINO_ASSERT(argc == 3, "Usage: ", argv[0], " <MODEL_DIR> '<PROMPT>'");

    const std::string models_path = argv[1], prompt = argv[2];
    const std::string device = "CPU";  // GPU, NPU can be used as well

    auto txt2img_callback = [](size_t iter_num, ov::Tensor& latents) -> bool {
        std::cout << "txt2img_callback " << iter_num << latents.get_shape() << std::endl;
        std::cout << "main data " << std::endl;
        for (int i = 0; i < 10; ++i)
            std::cout << latents.data<float>()[i] << " ";
        std::cout << std::endl;
        if (iter_num == 3)
            return true;
        return false;
    };

    ov::genai::Text2ImagePipeline pipe(models_path, device);
    ov::Tensor image = pipe.generate(prompt,
        ov::genai::width(512),
        ov::genai::height(512),
        ov::genai::num_inference_steps(20),
        ov::genai::num_images_per_prompt(2)
        , ov::genai::callback(txt2img_callback)
        );
        // std::function<bool(size_t, ov::Tensor&)>

    // writes `num_images_per_prompt` images by pattern name
    imwrite("image_%d.bmp", image, true);

    return EXIT_SUCCESS;
} catch (const std::exception& error) {
    try {
        std::cerr << error.what() << '\n';
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
} catch (...) {
    try {
        std::cerr << "Non-exception object thrown\n";
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
}

File: samples/python/text2image/main.py
#!/usr/bin/env python3
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import openvino
import openvino_genai
import torch
import numpy as np

from PIL import Image

class Generator(openvino_genai.Generator):
    def __init__(self, seed):
        openvino_genai.Generator.__init__(self)
        self.generator = torch.Generator(device='cpu').manual_seed(seed)

    def next(self):
        print('call next')
        return torch.randn(1, generator=self.generator, dtype=torch.float32).item()

    def randn_tensor(self, shape: openvino.Shape):
        torch_tensor = torch.randn(list(shape), generator=self.generator, dtype=torch.float32)
        print('generate tensor')
        return openvino.Tensor(torch_tensor.numpy())


def read_image(path: str) -> openvino.Tensor:
    pic = Image.open(path).convert("RGB")
    print(f"pic.size = {pic.size}")
    image_data = np.array(pic.getdata()).reshape(1, pic.size[1], pic.size[0], 3).astype(np.uint8)
    return openvino.Tensor(image_data)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('model_dir')
    parser.add_argument('prompt')
    parser.add_argument('image')
    parser.add_argument('mask')
    args = parser.parse_args()

    device = 'CPU'  # GPU can be used as well
    pipe = openvino_genai.Image2ImagePipeline(args.model_dir, device)

    initial_image = read_image(args.image)
    _, H, W, _ = list(initial_image.shape)
    H = H // 2
    W = W // 2
    mask_image = read_image(args.mask)

    image_tensor = pipe.generate(
        args.prompt,
        initial_image,
        mask_image,
        width=W,
        height=H,
        strength=0.8,
        generator=Generator(42)
    )

    image = Image.fromarray(image_tensor.data[0])
    image.save("/home/devuser/ilavreno/openvino.genai/genai_image.bmp")


if '__main__' == __name__:
    main()
